earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp

```diff
@@ -278,16 +278,105 @@ class StateUpdater::SetStateVisitor : public default_dfs_visitor {
     }
 };
 
+
+class StateUpdater::SetBlockingStateVisitor : public default_dfs_visitor {
+  private:
+    using RecalcSet = std::unordered_set<SharedString>;
+
+    StateUpdater & updater;
+    DependentStateTree & tree;
+    const AssetDefs::State newState;
+    // This is a shared_ptr because the visitor will be copied several times
+    // during the course of the traversal.
+    const std::shared_ptr<RecalcSet> hasBlockingInputs;
+    const std::shared_ptr<RecalcSet> hasBlockingChildren;
+
+    void SetState(
+        DependentStateTreeVertexDescriptor vertex,
+        AssetDefs::State newState,
+        const WaitingFor & waitingFor) const {
+      SharedString name = tree[vertex].name;
+      AssetDefs::State oldState = tree[vertex].state;
+      if (newState != oldState) {
+        notify(NFY_PROGRESS, "Setting state of '%s' from '%s' to '%s'",
+              name.toString().c_str(), ToString(oldState).c_str(), ToString(newState).c_str());
+        auto version = updater.storageManager->GetMutable(name);
+        if (version) {
+            updater.SetVersionStateAndRunHandlers(version, newState, waitingFor);
+
+          // Get the new state directly from the asset version since it may be
+          // different from the passed-in state
+          auto & data = tree[vertex];
+          data.state = version->state;
+
+          if (AssetDefs::Canceled == newState || AssetDefs::Blocked == newState || AssetDefs::Failed == newState){
```

Comment:
There's a fair amount of duplication between the two SetState functions. It would be nice if we could combine them somehow. I think this if statement is the only significant difference. I guess you could change the other SetState function to have different sets for children and inputs and then check both of them in the corresponding finish_vertex function. Then I think you could pull the common functionality into a parent class and remove the duplicate function.

Just thinking out loud. I always try to get rid of as much duplication as possible.

Reply:
Are you talking about pulling both `SetState` and `finish_vertex` into a base class, or just `SetState`?

Reply:
Just SetState. It would be great if we could get the common parts of finish_vertex into a base class as well, but that seems harder.

Reply:
I created the base class and pulled SetState into it. I opted to make it a template method and let each derived class keep its separate sets. They could have used sets in the base class, but I like that the names are more explicit and a littler clearer this way. I could be talked out of that if there's a strong feeling to the contrary.